### PR TITLE
Core: add apply_rule to rule_grouping, closes #3461

### LINKED ISSUE #3461: add sharded apply_rule function to rule_grouping
Motivation
----------
to address memory scaling problems in add_rule and inject_rule create equivalent of  current double call of **__resolve_did_to_locks_and_replicas** and **__create_locks_replicas_transfers**
that will process the dataset in multiple bunches of fixed max size

Modification
------------

### PR DESCRIPTION
**Pull request by lgoossen**
Core: add apply_rule to rule_grouping, closes #3461

### CODE REVIEW

**Comment by lgoossen** [comment]
for code review

File: lib/rucio/core/rule_grouping.py [hunk 1]
```diff
@@ -1096,3 +1103,330 @@ def __set_replica_unavailable(replica, session=None):
...
+
+    :param did:          the did object
+    :param rule:         the rule object
+    :param rses:         target rses_ids
+    :param source_rses:  source rses_ids
+    :param rseselector:  the rseselector object
+    :param session:      the database session in use
+    """
+
+    max_partition_size = core_config_get('rules', 'apply_rule_max_partition_size', default=2000)  # process dataset files in bunches of max this size
```
Comment (bari12) [inline lib/rucio/core/rule_grouping.py]: Add `session=session` to the method call

File: lib/rucio/core/rule_grouping.py [hunk 2]
```diff
@@ -1096,3 +1103,330 @@ def __set_replica_unavailable(replica, session=None):
...
+        used_rse_ids = []   # for NONE grouping keep track of actual used RSEs
+
+        if rule.grouping == RuleGrouping.ALL:
+            # calculate target RSEs
+            nbytes = 0
+            rse_coverage = {}
+            # simply loop over child datasets
+            # this is an approximation because ignoring the possibility of file overlap
+            for ds_scope, ds_name in datasets:
+                ds = rucio.core.did.get_did(scope=ds_scope, name=ds_name, dynamic=True)  # this will be retrieved again later on -> could be optimized
```
Comment (bari12) [inline lib/rucio/core/rule_grouping.py]: Add `session=session` to the method call

File: lib/rucio/core/rule_grouping.py [hunk 3]
```diff
@@ -1096,3 +1103,330 @@ def __set_replica_unavailable(replica, session=None):
...
+                nbytes += ds['bytes']
+                one_rse_coverage = rucio.core.replica.get_RSEcoverage_of_dataset(scope=ds_scope, name=ds_name)
```
Comment (bari12) [inline lib/rucio/core/rule_grouping.py]: This method is not part of the module yet?
Add `session=session` to the method call

File: lib/rucio/core/rule_grouping.py [hunk 4]
```diff
@@ -1096,3 +1103,330 @@ def __set_replica_unavailable(replica, session=None):
...
+            preferred_rse_ids = [x[0] for x in sorted(rse_coverage.items(), key=lambda tup: tup[1], reverse=True)]
+            # prnt(preferred_rse_ids)
+            rse_tuples = rseselector.select_rse(size=nbytes, preferred_rse_ids=preferred_rse_ids,
+                                                prioritize_order_over_weight=True, existing_rse_size=rse_coverage)
+            # prnt(rse_tuples)
+
+        for ds_scope, ds_name in datasets:
+            # prnt(('processing dataset ',ds_scope, ds_name))
+            #
+            ds = rucio.core.did.get_did(scope=ds_scope, name=ds_name, dynamic=True)
```
Comment (bari12) [inline lib/rucio/core/rule_grouping.py]: Add `session=session` to the method call

File: lib/rucio/core/rule_grouping.py [hunk 5]
```diff
@@ -1096,3 +1103,330 @@ def __set_replica_unavailable(replica, session=None):
...
+            ds_open = ds['open']
+            # prnt(ds)
+
+            # calculate number of partitions based on nr of files
+            npartitions = int(ds_length / max_partition_size) + 1
+            # prnt(npartitions)
+
+            if rule.grouping == RuleGrouping.DATASET:
+                # calculate target RSEs
+                rse_coverage = rucio.core.replica.get_RSEcoverage_of_dataset(scope=ds_scope, name=ds_name)
```
Comment (bari12) [inline lib/rucio/core/rule_grouping.py]: Add `session=session` to the method call

File: lib/rucio/core/rule_grouping.py (hunk 3 above)
Comment (bari12) [reply lib/rucio/core/rule_grouping.py]: Apologies, the method was added to the module in an earlier PR.

**Review by bari12 — APPROVED** [review]

**Pull request merged**